rust-mode.el

```diff
@@ -1332,29 +1332,49 @@ This is written mainly to be used as `end-of-defun-function' for Rust."
 ;; Formatting using rustfmt
 (defun rust--format-call (buf)
   "Format BUF using rustfmt."
-  (with-current-buffer (get-buffer-create "*rustfmt*")
-    (erase-buffer)
-    (insert-buffer-substring buf)
-    (let* ((tmpf (make-temp-file "rustfmt"))
-           (ret (call-process-region (point-min) (point-max) rust-rustfmt-bin
-                                     t `(t ,tmpf) nil)))
-      (unwind-protect
+  (let ((fmt-buffer (get-buffer-create "*rustfmt*")))
+    (with-current-buffer fmt-buffer
+      (let ((buffer-read-only nil))
+        (erase-buffer)
+        (insert-buffer-substring buf)
+        (let ((ret (shell-command-on-region (point-min) (point-max)
+                                            rust-rustfmt-bin fmt-buffer
+                                            nil nil t)))
           (cond
            ((zerop ret)
             (if (not (string= (buffer-string)
                               (with-current-buffer buf (buffer-string))))
                 (copy-to-buffer buf (point-min) (point-max)))
             (kill-buffer))
            ((= ret 3)
-            (if (not (string= (buffer-string)
-                              (with-current-buffer buf (buffer-string))))
-                (copy-to-buffer buf (point-min) (point-max)))
-            (erase-buffer)
-            (insert-file-contents tmpf)
-            (error "Rustfmt could not format some lines, see *rustfmt* buffer for details"))
+            (let ((path (buffer-file-name buf))
+                  (line-error "error: line exceeded maximum width[[:ascii:]]+")
+                  buf-string)
+              (copy-to-buffer buf (point-min) (point-max))
+              (with-current-buffer buf
+                (while (re-search-forward line-error nil t)
+                  (replace-match ""))
+                (setq buf-string (buffer-string)))
+              (goto-char (point-min))
+              (save-excursion
+                (while (re-search-forward buf-string nil t)
```

Comment:
I don't understand what this loop is for.  Could you add a comment explaining it?

Reply:
This removes the actual file contents from the compilation buffer, but you're right that can be done better. IMO rustfmt shouldn't format files if there are any kind of errors. I mean I have to take care of them anyway as I don't want an error message every time I use rustfmt. Hopefully that will be changed when it's more mature.

I can understand if you don't want to merge it, as it won't work correctly in every situation. I can try to make this more sound, but I can also wait until there's an improvement on the rustfmt side and give it another try.

Reply:
I know nothing, but I'm curious why we can't use [`compilation-start`](https://github.com/emacs-mirror/emacs/blob/master/lisp/progmodes/compile.el#L1633)?

Reply:
I think `compilation-start` doesn't allow piping through stdin.